packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs

```diff
@@ -0,0 +1,131 @@
+//! Implementation of [`ApplyTest`] for [`ContentFilter`]. This module contains the implementation details of
+//! filtering/querying a component and its props.
+
+use crate::{graph_node::GraphNode, props::PropProfile, DocumentModel};
+
+use super::{ApplyTest, Cond, PropValue};
+
+/// Filters used to filter content (props/components/strings) from the structure graph.
+#[derive(Debug, Clone)]
+pub enum ContentFilter {
+    /// Match a component by type name.
+    IsType(&'static str),
+    /// Match components that have a prop matching a profile;
+    /// The value of the matched prop is not considered.
+    HasPropMatchingProfile(PropProfile),
+    /// Match components that have a prop matching a profile;
+    /// The value of the matched prop is tested against `Cond` and only matches
+    /// if `Cond` is true.
+    HasPropMatchingProfileAndCondition(PropProfile, Cond<PropValue>),
+    /// Match content is a component.
+    IsComponent,
+    /// Match content is a string.
+    IsString,
+}
+
+/// Information needed to determine of `node` matches a given filter condition.
+#[derive(Debug, Clone)]
+pub struct FilterData<'a> {
+    pub node: GraphNode,
+    pub origin: GraphNode,
+    pub document_model: &'a DocumentModel,
+}
+
+impl ApplyTest<FilterData<'_>, GraphNode> for ContentFilter {
+    fn apply_test(&self, data: &FilterData) -> bool {
+        // Warning: Only use `get_prop_unchecked` for getting a prop in this function.
+        // This is so that we possibly catch errors a little earlier if a prop that is not
+        // set in the dependency graph is accessed.
+        let node = data.node;
+        let document_model = data.document_model;
+        let origin = data.origin;
+
+        match self {
+            ContentFilter::IsType(type_name) => match node {
+                GraphNode::Component(_) => {
+                    let component_type = document_model.get_component_type(node);
+                    component_type.eq(type_name)
+                }
+                _ => false,
+            },
+            ContentFilter::HasPropMatchingProfile(profile) => match node {
+                GraphNode::Component(component_idx) => {
+                    let prop =
+                        document_model.get_component_prop_by_profile(component_idx, &[*profile]);
+                    prop.is_some()
+                }
+                _ => {
+                    // `PropProfile::String` is the only profile that non-component nodes can match.
```

Comment:
Maybe clearer if this were in a `GraphNode::String(_)` match arm?

If we have a `GraphNode::String(_)`, it should return true if we have a `String` or  `LiteralString` prop profile. I don't have a test checking this as there is no longer a component using `LiteralString`. Is this an argument for removing `LiteralString` for now?

Reply:
If we remove it, will that mean we can no longer force certain attrs to be defined with an absence of references?

Reply:
Right. That's a good reason to keep `LiteralString`, otherwise we can't distinguish a reference turned text from a string.

Reply:
A patch:
```
diff --git a/packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs b/packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs
index d3ae1cc6f..f08cec7e0 100644
--- a/packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs
+++ b/packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/content_filter.rs
@@ -55,14 +55,9 @@ impl ApplyTest<FilterData<'_>, GraphNode> for ContentFilter {
                     prop.is_some()
                 }
                 GraphNode::String(_) => {
-                    // `PropProfile::String` is the only profile that non-component nodes can match.
-                    // We special case it here.
-                    if matches!(profile, PropProfile::String)
-                        && matches!(node, GraphNode::String(_))
-                    {
-                        return true;
-                    }
-                    false
+                    // `PropProfile::String` and `PropProfile::LiteralString` are the only profiles that non-component nodes can match.
+                    // We special case them here.
+                    matches!(profile, PropProfile::String | PropProfile::LiteralString)
                 }
                 _ => unreachable!("Only components and strings can match profiles"),
             },
```